crates/isobmff/derive/src/lib.rs

```diff
@@ -0,0 +1,698 @@
+            read_field(field, &crate_path)
+        };
+
+        fields_in_self.push(field_name.to_token_stream());
+        field_parsers.push(quote! {
+            let #field_name = #read_field;
+        });
+
+        match (field.repeated, &field.from) {
+            (true, None) => {
+                field_serializers.push(quote! {
+                    for item in &self.#field_name {
+                        #crate_path::reexports::scuffle_bytes_util::zero_copy::Serialize::serialize(item, &mut writer)?;
+                    }
+                });
+            }
+            (true, Some(from_ty)) => {
+                field_serializers.push(quote! {
+                    for item in &self.#field_name {
+                        #crate_path::reexports::scuffle_bytes_util::zero_copy::Serialize::serialize(&::std::convert::Into::<#from_ty>::into(*item), &mut writer)?;
+                    }
+                });
+            }
+            (false, None) => {
+                field_serializers.push(quote! {
+                    #crate_path::reexports::scuffle_bytes_util::zero_copy::Serialize::serialize(&self.#field_name, &mut writer)?;
+                });
+            }
+            (false, Some(from_ty)) => {
+                field_serializers.push(quote! {
+                    #crate_path::reexports::scuffle_bytes_util::zero_copy::Serialize::serialize(&::std::convert::Into::<#from_ty>::into(self.#field_name), &mut writer)?;
+                });
+            }
+        }
+    }
+
+    let collect_boxes = fields.iter().any(|f| f.nested_box.is_some());
+
+    let box_parser = if collect_boxes {
+        Some(nested_box_parser(fields.iter(), &crate_path))
+    } else {
+        None
+    };
+
+    for (field, nested) in fields.iter().filter_map(|f| f.nested_box.map(|n| (f, n))) {
+        let field_name = field.ident.clone().expect("unreachable: only named fields supported");
+        let field_name_str = field_name.to_string();
+
+        match nested {
+            IsoBoxFieldNestedBox::Single => {
+                fields_in_self.push(quote! {
+                    #field_name: ::std::option::Option::ok_or(#field_name, ::std::io::Error::new(::std::io::ErrorKind::InvalidData, format!("{} not found", #field_name_str)))?
+                });
+                field_serializers.push(quote! {
+                    #crate_path::reexports::scuffle_bytes_util::zero_copy::Serialize::serialize(&self.#field_name, &mut writer)?;
+                });
+            }
+            IsoBoxFieldNestedBox::Collect | IsoBoxFieldNestedBox::CollectUnknown => {
+                fields_in_self.push(field_name.to_token_stream());
+                field_serializers.push(quote! {
+                    #[allow(for_loops_over_fallibles)]
+                    for item in &self.#field_name {
+                        #crate_path::reexports::scuffle_bytes_util::zero_copy::Serialize::serialize(item, &mut writer)?;
+                    }
+                });
+            }
+        }
```

Comment:
I suspect this can be simplified to be handled by the trait objects instead of generating an iterator.